### ISSUE TITLE
Skinning system

### ISSUE THREAD

**Issue by vchelaru** [vchelaru/Gum#13]
@bjorn says:

I'm missing explicit support for skins. For most use-cases, I may be able to put a child component at the bottom called "skin" and deal with that in my export script, but our skin system as well as that of Flex has a feature that is not available this way.

This is the ability to define the content area of a component, which affects the layout positions of its children as well as when the children get drawn in relation to other visual parts of the skin. So, you can have a skin that adds a border and decreases the area available for the children, or a skin that adds visuals both behind and in front of the children of the skinned component.

**Comment by vchelaru** [comment]
I think the hardest part of this is a skinning system which allows for the creation of brand new components. I'm not sure how to handle this.

I have thought about it a bit and I think what might be good is to have a section for global variables. For example, a global variable could be referencing a folder for all textures. Then variables could reference global variables in part or in whole.

This would require simple "operations" for variables as well, if the skinning is to be done with different source folders. For example, if an app has two root folders called Skin1 and Skin2, then there would be a global variable called RootSkinFolder and the variable for texture might be:

{global::RootSkinFolder} + Buttons/Button.png

@bjorn thoughts?

**Comment by bjorn** [comment]
@vchelaru The thing about global variables and skin folders seem to have little to do with my request. I'm not talking about applying a global skin to change the look of the UI components. Rather, this request is only about a system where the behavior and the appearance of a component can be separated.

In the simple case, I have a Button component that implements click behavior and related state handling, and in Gum I would be able to instantiate this component multiple times, but specify a different skin on each button instance because buttons rarely look the same everywhere. Because in this case, the appearance of the button is entirely defined by its skin, this use-case is already covered by Gum by simply placing the "skin" instance as a child of the "button" instance. The only downside in this simple case is that the default size of the button is not taken from its skin, so it needs to be set explicitly on the button instance.

The cases that are not currently supported are basically things like scroll areas or windows, or skinnable containers in general. The biggest limitations being the ones I listed above, namely that the skin can't place things both in front and behind the children and that it can't affect the area where the children get placed.

**Comment by vchelaru** [comment]
I am closing this issue out. What is being asked for is basically something similar to Behaviors + Forms which are the current pattern for something like this. At least in the XNA-like runtimes, you can do

```cs
var button = new Button();
```

and elsewhere specify default implementations for Button or whatever Forms control you want.

The original request was using a custom runtime rather than actual Gum runtimes, and the original runtime is no longer maintained (afaik) so if the current Forms solution is not suitable for modern projects, we can open up a new issue.

**Comment by bjorn** [comment]
Yeah that seems fine. Indeed the Lua-based runtime I was exporting to unfortunately was never used again after the release of Champions of Anteria. As far as I remember I did some kind of workaround in our export plugin to have it hook the right behaviors onto our controls.

I can't believe we're 10 years on now. Great to see you're still maintaining Gum! It seems to be finally getting the attention it deserves!
